mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseToStdlib.cpp

```diff
@@ -1456,45 +1467,130 @@ struct XdlopsGemmV2RewritePattern : public OpRewritePattern<XdlopsGemmV2Op> {
-            ValueRange{offset});
-        argB = innerLoopb.create<vector::TransferReadOp>(
-            loc, argType.template cast<VectorType>(), op.bufferB(),
-            ValueRange{offset});
+      assert(argTypeVectorLength == k_base);
+
+      if (KPack == 1) {
+        // When KPack == 1, we are dealing with memref<T>.
+        // Use a combination of inner and outer offset to figure out the actual
+        // offset from vgpr
+        Value offset = innerLoopb.create<AddIOp>(
+            loc, innerLoopb.create<MulIOp>(loc, outerLoopiv, KBaseConstantOp),
+            innerLoopb.create<MulIOp>(loc, innerLoopiv, KBaseConstantOp));
+
+        if (k_base == 1) {
+          // xdlops needs only 1 element, load directly from buffer.
+          argA = innerLoopb.create<memref::LoadOp>(loc, argType, op.bufferA(),
+                                                   ValueRange{offset});
+          argB = innerLoopb.create<memref::LoadOp>(loc, argType, op.bufferB(),
+                                                   ValueRange{offset});
+        } else {
+          // k_base > 1, use transferRead to load a vector length equivalent
+          // with a xdlops argument.
+          argA = innerLoopb.create<vector::TransferReadOp>(
+              loc, argType.cast<VectorType>(), op.bufferA(),
+              ValueRange{offset});
+          argB = innerLoopb.create<vector::TransferReadOp>(
+              loc, argType.cast<VectorType>(), op.bufferB(),
+              ValueRange{offset});
+        }
       } else {
-        argA = innerLoopb.create<memref::LoadOp>(loc, argType, op.bufferA(),
-                                                 ValueRange{offset});
-        argB = innerLoopb.create<memref::LoadOp>(loc, argType, op.bufferB(),
-                                                 ValueRange{offset});
+        // This lambda create an xdlops argument from the vgpr buffer.
+        // Each time in the nested loop, it first loads a vector of size
+        // kpack guarateed to be a multiple of the size of the xdlops
+        // argument. Then it construct the xdlops argument by extracting
+        // elements from the kpack vector.
+        auto constructXdlopsArg = [&outerLoopb, &innerLoopb,
+                                   &loc](Value &buffer, Value &outerLoopiv,
+                                         Value &innerLoopiv, Type &argType,
+                                         int64_t k_base) -> Value {
+          Value arg = createZeroConstantOp(innerLoopb, loc, argType);
+
+          Value argAWide = innerLoopb.create<memref::LoadOp>(
+              loc,
+              buffer.getType()
+                  .cast<MemRefType>()
+                  .getElementType()
+                  .cast<VectorType>(),
+              buffer, ValueRange{outerLoopiv});
+
+          for (int64_t i = 0; i < k_base; ++i) {
+            Value iterOp = innerLoopb.create<ConstantIndexOp>(loc, i);
+            Value newOffset =
```

Comment:
This is `miopen::ExtractSliceOp`

Reply:
Done in c5ae273. Nice! This just make the lambda trivial!